src/coreclr/jit/morph.cpp

```diff
@@ -11053,6 +11053,60 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
     return op1;
 }
 
+//------------------------------------------------------------------------------
+// fgMorphOpCast : Try to simplify "(T)x op (T)y" to "(T)(x op y)".
+// https://github.com/dotnet/runtime/issues/13816
+//
+// Arguments:
+//       tree - node to fold
+//
+// return value:
+//       A folded GenTree* instance, or the same GenTree* if it couldn't be folded
+//
+
+GenTree* Compiler::fgMorphOpCast(GenTreeOp* tree)
+{
+    assert(varTypeIsIntegralOrI(tree->TypeGet()));
+    assert(tree->OperIs(GT_OR, GT_AND, GT_XOR));
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+    genTreeOps oper = tree->OperGet();
+
+    // see whether both ops are casts, with matching to and from types
+    if (op1->OperIs(GT_CAST) && op2->OperIs(GT_CAST))
+    {
+        var_types fromType = op1->CastFromType();
+        var_types toType = op1->CastToType();
+        GenTreeFlags gtf_unsigned = op1->gtFlags & GTF_UNSIGNED;
+
+        if (op2->CastFromType() != fromType || op2->CastToType() != toType || (op2->gtFlags & GTF_UNSIGNED) != gtf_unsigned)
+        {
+            return tree;
+        }
+
+        //     tree             CAST
+        //     /   \             |
+        //   CAST  CAST   ==>   tree
+        //    |     |          /   \
+        //   op1   op2        op1  op2
+
+        tree->gtOp1 = op1->AsCast()->CastOp();
+        tree->gtOp2 = op2->AsCast()->CastOp();
+        tree->gtType = fromType;
+
+        GenTree* result = gtNewCastNode(toType, tree, gtf_unsigned != 0, toType);
+
+        DEBUG_DESTROY_NODE(op1);
```

Comment:
Since such transformations are usually invoked from different phases, not just global morph - we need to make sure we're not folding active CSE candidates, so both both op1 and opt2 have to checked via `gtIsActiveCSE_Candidate`

Reply:
It is probably wise to just `fgGlobalMorph` it away, we cannot easily regenerate the VN for the operator node.

Reply:
Just wanna make sure I understand what you're saying (for the benefit of my own learning!):

1. We don't wanna fold CSE candidates (because then they wouldn't be CSE candidates any more)
2. CSE happens after global morphing, so wrapping this in `if (fgGlobalMorph)` fixes (1)
3. We don't need do to this outside of `fgGlobalMorph` anyway

Is that logic correct?

Reply:
> We don't wanna fold CSE candidates (because then they wouldn't be CSE candidates any more)

We also don't want to fold because CSE holds links (`GenTree*`s) to CSE candidates and will try to find them in the tree after morphing. Failure is fatal.

> CSE happens after global morphing, so wrapping this in if (fgGlobalMorph) fixes (1)

Yep.

> We don't need do to this outside of `fgGlobalMorph` anyway

It is not that we don't need, we cannot, easily at least. The global value numbering pass assigns VNs to all trees, and morph must preserve them (failure mode here is silent bad code generation). This transform can retype a `TYP_INT` operator node to `TYP_LONG`, thus altering the value it produces, necessitating a VN update. But such an update is not trivial, since it requires calling `fgValueNumberTree`, which is not tested to work outside the global value numbering pass (indeed, with the recent changes it will fail sometimes unless you are careful). We thus shy away from updating VNs in these more "complex" cases, instead disabling the optimization for the optimizer phases. The two cases where we do prefer to update VNs are: 1) a constant changes its value (`fgUpdateConstTreeValueNumber` is the method to call in such a case), 2) the new tree can inherit its VN from some other tree.

Reply:
Thank you very much for the detailed answer. This makes a lot of sense.

cdec019d463276d2546e47e51d6894e951f0a7b7